source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst

```diff
@@ -1502,6 +1617,33 @@ Once the client has taken the lock it must do no I/O::
         client.condition.notifyAll()
         client.lock.release()
 
+    def compareTopologyVersion(tv1, tv2):
+        """Return -1 if tv1<tv2, 0 if tv1==tv2, 1 if tv1>tv2"""
+        if tv1 is None or tv2 is None:
+            # Assume greater.
+            return -1
+        pid1 = tv1['processId']
+        pid2 = tv2['processId']
+        if pid1 == pid2:
+            counter1 = tv1['counter']
+            counter2 = tv2['counter']
+            if counter1 == counter2:
+                return 0
+            elif counter1 < counter2:
+                return -1
+            else:
+                return 1
+        else:
+            # Assume greater.
+            return -1
+
+    def shouldHandleServerDescription(topologyDescription, server):
```

Comment:
I ran into some minor issues implementing this that I think could be corrected by being more explicit in our specification. Specifically, if `topologyDescription` or `server` in this case are null, then you are accessing invalid properties on those types below. In JavaScript we can be a little clever:
```
function shouldHandleServerDescription(previousServerDescription, incomingServerDescription) {
  const previousTopologyVersion =
    previousServerDescription && previousServerDescription.topologyVersion;
  const incomingTopologyVersion =
    incomingServerDescription && incomingServerDescription.topologyVersion;

  return compareTopologyVersion(previousTopologyVersion, incomingTopologyVersion) <= 0;
}
```
But I almost prefer being insanely verbose in documentation that all drivers will have to follow:
 - If (topologyDescription == null || server == null) return true
 - If you can't assume that a server is valid before you pass it in, then check that `topologyDescription.hasServer(server.address)` before you use `currentServer`
 - check that `currentServer` has a `topologyVersion` before using that

Feels like a nitpick, but I think it pays to be verbose in these situations.

Reply:
How could topologyDescription or `incomingServerDescription` ever be null? Can you point me to where the spec says that could be the case?

Reply:
oops, I had placed my logic before the pseudocode to verify that the incoming description was a known server. Disregard.